src/actix/api/snapshot_api.rs

```diff
@@ -211,6 +211,8 @@ async fn upload_snapshot(
             &collection.name,
             snapshot_recover,
             access,
+            // No need to limit to snapshot directory, only uploaded file is used
+            false,
```

Comment:
Is it necessary to set this to `false`? Couldn't we set it to `true` just to be safe, since `do_save_uploaded_snapshot` already tries to force the file to be in the snapshot directory? https://github.com/qdrant/qdrant/blob/5c29cad76edc4fe33882d385c02633352666d67e/src/actix/api/snapshot_api.rs#L110

Currently we use string validation to block relative paths in the filename in the `multipart/form-data` request body and collection name, which could mean slight changes could make the vuln come back, so I like the idea of checking the path twice if it doesn't break other stuff

Reply:
> Is it necessary to set this to false? Couldn't we set it to true just to be safe, since do_save_uploaded_snapshot already tries to force the file to be in the snapshot directory? 

Yes that is required, in the current implementation. When uploading it might store snapshots in a different directory if the user configured it that way, in which case the check would fail.

You're right that we use string-ish validation, but we can not really represent file paths in a different way. To add a bit of nuance here, we use [`Path`](https://doc.rust-lang.org/std/path/struct.Path.html) from the standard library here. I believe we can assume this to be safe for what we're trying to do here - strip everything but a file name from an arbitrary path. Note that if this step fails, we pick a random UUID.

The only solution I can think of is to use something like [`cap-std`](https://docs.rs/cap-std/latest/cap_std/) which enforces directory namespacing. However, it is not compatible with our current snapshot logic. I did attempt to integrate this before, but had to decide it was not worth the effort.

So what's your opinion on this after the above?

Reply:
> When uploading it might store snapshots in a different directory if the user configured it that way, in which case the check would fail.

If the user configured a different snapshots directory, couldn't we read that configuration and enforce the file is being saved there? Or am I misunderstanding?

Overall this PR looks good, just wanted to ask if flipping that boolean could make us a little safer for free. Sounds like it's not that simple :)

Reply:
We don't move the file, but we do directly store it in the snapshots directory now.

IIRC we documented it differently, like I described above. We either have to update documentation, or store in the proper directory.

Reply:
Am I missing something?

- https://github.com/qdrant/qdrant/blob/dev/src/actix/api/snapshot_api.rs#L196-L197
- https://github.com/qdrant/qdrant/blob/dev/src/actix/api/snapshot_api.rs#L89-L112

Reply:
Also:
- https://github.com/qdrant/qdrant/blob/dev/src/actix/mod.rs#L136
- https://github.com/qdrant/qdrant/blob/dev/src/actix/mod.rs#L66-L69
- https://github.com/qdrant/qdrant/blob/master/lib/storage/src/content_manager/toc/temp_directories.rs#L122-L128

Reply:
Yes, but this whole thread is about `upload_snapshot`?